// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: lorawan-stack/api/contact_info.proto

package api

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/golang/protobuf/ptypes"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = ptypes.DynamicAny{}
)

// define the regex for a UUID once up-front
var _contact_info_uuidPattern = regexp.MustCompile("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$")

// Validate checks the field values on ContactInfo with the rules defined in
// the proto definition for this message. If any rules are violated, an error
// is returned.
func (m *ContactInfo) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for ContactType

	// no validation rules for ContactMethod

	// no validation rules for Value

	// no validation rules for Public

	if v, ok := interface{}(m.GetValidatedAt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ContactInfoValidationError{
				field:  "ValidatedAt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// ContactInfoValidationError is the validation error returned by
// ContactInfo.Validate if the designated constraints aren't met.
type ContactInfoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactInfoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactInfoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactInfoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactInfoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactInfoValidationError) ErrorName() string { return "ContactInfoValidationError" }

// Error satisfies the builtin error interface
func (e ContactInfoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactInfo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactInfoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactInfoValidationError{}

// Validate checks the field values on ContactInfoValidation with the rules
// defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *ContactInfoValidation) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Id

	// no validation rules for Token

	if v, ok := interface{}(m.GetEntity()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ContactInfoValidationValidationError{
				field:  "Entity",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	for idx, item := range m.GetContactInfo() {
		_, _ = idx, item

		if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ContactInfoValidationValidationError{
					field:  fmt.Sprintf("ContactInfo[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if v, ok := interface{}(m.GetCreatedAt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ContactInfoValidationValidationError{
				field:  "CreatedAt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if v, ok := interface{}(m.GetExpiresAt()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ContactInfoValidationValidationError{
				field:  "ExpiresAt",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// ContactInfoValidationValidationError is the validation error returned by
// ContactInfoValidation.Validate if the designated constraints aren't met.
type ContactInfoValidationValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ContactInfoValidationValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ContactInfoValidationValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ContactInfoValidationValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ContactInfoValidationValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ContactInfoValidationValidationError) ErrorName() string {
	return "ContactInfoValidationValidationError"
}

// Error satisfies the builtin error interface
func (e ContactInfoValidationValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sContactInfoValidation.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ContactInfoValidationValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ContactInfoValidationValidationError{}
